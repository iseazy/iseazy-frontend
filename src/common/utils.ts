export function getFormattedTimeDifference(fromDate: number, toDate: number = Date.now()): string {
    const millis = toDate -fromDate;
    const minutes = Math.floor(millis / 60000);
    const seconds = ((millis % 60000) / 1000).toFixed(0);
    let gameDuration = '';
    
    if (minutes > 0) {
        gameDuration += `${minutes}min `;
    }

    return `${gameDuration}${seconds}s`;
}

export function shuffleList<T>(items: Array<T>): Array<T> {
    return items.sort(() => Math.random() - 0.5);
}
