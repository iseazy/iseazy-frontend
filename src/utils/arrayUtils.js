import _ from 'lodash';

const shuffleItems = (array) => !_.isNil(array) && _.shuffle(array);

const findItemById = (items, itemId) =>
  items.find((item) => item.id === itemId);

const updateItemByIndex = (items, itemId, props) =>
  items.map((item) => {
    if (item.id === itemId) {
      return {
        ...item,
        ...props,
      };
    } else {
      return item;
    }
  });

const updateItemsByIndex = (items, idsArray, props) =>
  items.map((item) => {
    if (idsArray.includes(item.id)) {
      return {
        ...item,
        ...props,
      };
    } else {
      return item;
    }
  });

export { shuffleItems, findItemById, updateItemByIndex, updateItemsByIndex };
