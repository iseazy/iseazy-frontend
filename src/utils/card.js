export const createCard = (src, sharedId) => ({
    sharedId,
    src,
    flipped: false,
    clicked: false,
});

export const assignCardId = (card, id) => ({
    ...card,
    id
});

export const flipCardById = (id) => (card) => card.id === id ? ({
    ...card,
    flipped: true,
    clicked: true
}) : card;

export const unflipClickedCard = (card) => !!card.clicked ? ({
    ...card,
    flipped: false,
    clicked: false
}) : card;

export const unclickCard = (card) => ({
    ...card,
    clicked: false
});

export const isCardClicked = ({ clicked }) => !!clicked;

export const isCardNotClicked = ({ clicked }) => !clicked;

export const isCardFlipped = ({ flipped }) => !!flipped;

export const hasCardSharedId = (sharedId) => (card) => card.sharedId === sharedId;
