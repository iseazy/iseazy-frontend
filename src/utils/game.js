import { GAME_STATE } from '../constants';

export const isGameEnd = ({ gameState }) => gameState === GAME_STATE.END;

export const isGameIdle = ({ gameState }) => gameState === GAME_STATE.IDLE;
