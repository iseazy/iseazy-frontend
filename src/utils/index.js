export {
    assignCardId,
    createCard,
    flipCardById,
    isCardClicked,
    isCardFlipped,
    isCardNotClicked,
    hasCardSharedId,
    unclickCard,
    unflipClickedCard
} from './card';

export {
    areClickedCardsPaired,
    createDeck,
    isDeckSolved
} from './deck';

export { isGameEnd, isGameIdle } from './game';

export { shuffleArray } from './suffle-array';

export { getFormattedTime } from './time';
