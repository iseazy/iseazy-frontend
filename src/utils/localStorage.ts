const SCORES_KEY = 'iseazy-scores';

export const saveScore = (score: number): void => {
  const storedValue = localStorage.getItem(SCORES_KEY)

  if (storedValue) {
    const scores: number[] = JSON.parse(storedValue);
    scores.push(score);

    scores.sort((a, b) => b - a);
    localStorage.setItem(SCORES_KEY, JSON.stringify(scores.slice(0, 3)));
  } else {
    localStorage.setItem(SCORES_KEY, JSON.stringify([score]));
  }
}

export const getScores = (): number[] => {
  const storedValue = localStorage.getItem(SCORES_KEY)

  if (!storedValue) {
    return [];
  }

  return JSON.parse(storedValue);
}