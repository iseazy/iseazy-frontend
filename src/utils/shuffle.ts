export function shuffle<T>(array: T[]): T[] {
  const newArray = [...array];
  const length = newArray.length;
  
  for (let i = 0; i < length; i++) {
    const randomIndex = Math.floor(Math.random() * length);
    
    const temp = newArray[i];
    newArray[i] = newArray[randomIndex];
    newArray[randomIndex] = temp;
  }
  
  return newArray;
}