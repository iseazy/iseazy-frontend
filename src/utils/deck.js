import { shuffleArray } from './suffle-array';
import { assignCardId, createCard, isCardFlipped, isCardNotClicked, hasCardSharedId } from './card';

export const createDeck = (source) => {
    const uniqueCards = source.map(createCard);
    const duplicatedCards = [ ...uniqueCards, ...uniqueCards ];
    const identifiedCards = duplicatedCards.map(assignCardId);

    return shuffleArray(identifiedCards);
};

export const areClickedCardsPaired = ([ { sharedId }, ...rest ]) => !!rest.length && rest.every(hasCardSharedId(sharedId))

export const isDeckSolved = (cards) => cards.every(isCardFlipped) && cards.every(isCardNotClicked);
