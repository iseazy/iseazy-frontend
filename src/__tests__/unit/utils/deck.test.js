import {
    createDeck,
    areClickedCardsPaired,
    isDeckSolved
} from '../../../utils/deck';
import {
    assignCardId,
    createCard,
    isCardFlipped,
    isCardNotClicked,
    hasCardSharedId
} from '../../../utils/card';
import { shuffleArray } from '../../../utils/suffle-array';

jest.mock('../../../utils/suffle-array', () => ({
    shuffleArray: jest.fn()
}));
jest.mock('../../../utils/card', () => ({
    assignCardId: jest.fn(),
    createCard: jest.fn(),
    isCardFlipped: jest.fn(),
    isCardNotClicked: jest.fn(),
    hasCardSharedId: jest.fn()
}));

const testSource = [
    'sourceA',
    'sourceB',
    'sourceC',
];

const testCards = [
    { sharedId: 1 },
    { sharedId: 2 },
    { sharedId: 3 },
    { sharedId: 4 },
    { sharedId: 5 },
]

describe('createDeck()', () => {
    beforeEach(() => {
        shuffleArray.mockClear().mockImplementation((a) => a);
        createCard.mockClear().mockImplementation((a) => ({ a }));
        assignCardId.mockClear().mockImplementation((a, i) => ({ ...a, i }));
    });

    it('should return a deck', () => {
        expect(createDeck(testSource)).toMatchSnapshot();
    });

    it('should shuffle the created deck', () => {
        createDeck(testSource);
        expect(shuffleArray).toHaveBeenCalledTimes(1);
    });

    it('should create as many cards as source elements', () => {
        createDeck(testSource);
        expect(createCard).toHaveBeenCalledTimes(testSource.length);
    });

    it('should have source elements twiced', () => {
        expect(createDeck(testSource).length).toBe(testSource.length * 2);
    });
});

describe('areClickedCardsPaired()', () => {
    beforeEach(() => {
        hasCardSharedId.mockReturnValue(() => true);
    });

    it('should return true when all cards have common sharedId', () => {
        expect(areClickedCardsPaired(testCards)).toBe(true);
    });

    it('should return false when any card has different sharedId', () => {
        hasCardSharedId.mockReturnValueOnce(() => false);
        expect(areClickedCardsPaired(testCards)).toBe(false);
    });

    it('should return false when given one card', () => {
        expect(areClickedCardsPaired([ { sharedId: 1} ])).toBe(false);
    });
});

describe('isDeckSolved()', () => {
    beforeEach(() => {
        isCardFlipped.mockReturnValue(true);
        isCardNotClicked.mockReturnValue(true);
    });

    it('should return true when all given cards are flipped and not clicked', () => {
        expect(isDeckSolved(testCards)).toBe(true);
    });

    it('should return false when any given card is not flipped', () => {
        isCardFlipped.mockReturnValueOnce(false);
        expect(isDeckSolved(testCards)).toBe(false);
    });

    it('should return false when any given card is clicked', () => {
        isCardNotClicked.mockReturnValueOnce(false);
        expect(isDeckSolved(testCards)).toBe(false);
    });
});
