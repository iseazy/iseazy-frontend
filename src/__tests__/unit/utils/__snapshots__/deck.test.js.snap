// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDeck() should return a deck 1`] = `
Array [
  Object {
    "a": "sourceA",
    "i": 0,
  },
  Object {
    "a": "sourceB",
    "i": 1,
  },
  Object {
    "a": "sourceC",
    "i": 2,
  },
  Object {
    "a": "sourceA",
    "i": 3,
  },
  Object {
    "a": "sourceB",
    "i": 4,
  },
  Object {
    "a": "sourceC",
    "i": 5,
  },
]
`;
